File: coin-change.rb
def coin_change(coins, amount)
  dp = Array.new(amount + 1, amount + 1)
  dp[0] = 0
  1.upto(amount) do |i|
    coins.each_with_index do |coin, j|
      if(coin <= i)
        dp[i] = [dp[i], dp[i-coin]+1].min
      end
    end
  end
  dp[amount] > amount ? -1 : dp[amount]
end

puts coin_change([1,2,5], 11)

File: coin-change.cpp
#include <vector>
#include <iostream>

using namespace std;

class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        int max = amount + 1;
        vector<int> dp(max, max);
        dp[0] = 0;
        for (int i = 1 ; i <= amount; i++) {
            for (int j = 0; j < coins.size(); j++) {
                int coin = coins[j];
                if(coin <= i) dp[i] = min(dp[i], dp[i-coin] + 1);
            }
        }
        return dp[amount] > amount ? -1 : dp[amount];
    }
};

int main()
{
    Solution s;
    vector<int> coins {1, 2, 5};
    cout << s.coinChange(coins, 11);
}
